### PR TITLE
main, win: create test runner in windows

### PR DESCRIPTION
**Pull request by refack**
For this case, I could do it simpler than libuv (since the runner prints it's own errors)
Fixes: #1

### CODE REVIEW

File: mini-test.gyp [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "targets": [ {
     "target_name": "mini-test",
-    "type": "<!(gypkg type)",
+    "type": "static_library",
```
Comment (indutny) [inline mini-test.gyp]: I'd rather keep it. Any point in dropping it?

File: src/spawn-win.c [hunk 2]
```diff
@@ -1,10 +1,75 @@
...
+#include <stdlib.h>
+#include <windows.h> 
+#include <strsafe.h>
+
 #include "include/mini/common.h"
 
+static char executable_path[BUFSIZE];
+
+int popen_complete(const char* cmd)
+{
```
Comment (indutny) [inline src/spawn-win.c]: Style, please put `{` on the same line as `popen_complete`.

Let's make `popen_complete` static.

File: include/mini/common.h [hunk 3]
```diff
@@ -6,11 +6,20 @@
 #include <string.h>
 #include <errno.h>
 
+#define BUFSIZE 4096
```
Comment (indutny) [inline include/mini/common.h]: I'd rather move `BUFSIZE` to some private file. This file is a public API.

File: include/mini/common.h [hunk 4]
```diff
@@ -6,11 +6,20 @@
 #include <string.h>
 #include <errno.h>
 
+#define BUFSIZE 4096 
+
+#ifdef _WIN32
+#include <Windows.h>
+#define ABORT ExitProcess(42)
+#else
+#define ABORT abort()
```
Comment (indutny) [inline include/mini/common.h:13]: There is no `abort` on Windows? How do they live?

File: src/spawn-win.c [hunk 5]
```diff
@@ -1,10 +1,75 @@
...
+  STARTUPINFO StartupInfo;
+  ZeroMemory(&StartupInfo, sizeof(StartupInfo));
+  StartupInfo.cb = sizeof(StartupInfo);
+
+  PROCESS_INFORMATION ProcessInformation;
+  ZeroMemory(&ProcessInformation, sizeof(ProcessInformation));
+
+  //  BOOL WINAPI CreateProcess(
```
Comment (indutny) [inline src/spawn-win.c]: I know it sounds lame, but let's use `/* */` comments here for consistency.

File: src/spawn-win.c [hunk 6]
```diff
@@ -1,10 +1,75 @@
...
+    NULL,                     // Use parent's environment block
+    NULL,                     // Use parent's starting directory 
+    &StartupInfo,             // Pointer to STARTUPINFO structure
+    &ProcessInformation       // Pointer to PROCESS_INFORMATION structure
+  );
+  CHECK_EQ(ret, TRUE, "CreateProcess failed");
+
+  // Wait until child process exits.
+  DWORD wait_ret = WaitForSingleObject(ProcessInformation.hProcess, INFINITE);
+  CHECK_EQ(wait_ret, WAIT_OBJECT_0, "Process ended strange");
```
Comment (indutny) [inline src/spawn-win.c:51]: Does it happen on non-zero exit code?

**Review by indutny — CHANGES_REQUESTED** [review]
Few nits, otherwise LGTM. Thank you!

File: mini-test.gyp (hunk 1 above)
Comment (refack) [reply mini-test.gyp]: No, local code leak.

Comment (refack) [reply mini-test.gyp]: I wanted to run "real" `GYP` to get Visual Studio project files 😞

File: src/spawn-win.c (hunk 2 above)
Comment (refack) [reply src/spawn-win.c]: Ack

File: src/spawn-win.c (hunk 5 above)
Comment (refack) [reply src/spawn-win.c]: NP

File: src/spawn-win.c (hunk 6 above)
Comment (refack) [reply src/spawn-win.c:51]: No, only on strage cases

Return code/value | Description |
----------------- | ----------- |
WAIT_ABANDONED 0x00000080L | The specified object is a mutex object that was not released by the thread that owned the mutex object before the owning thread terminated. Ownership of the mutex object is granted to the calling thread and the mutex state is set to nonsignaled. If the mutex was protecting persistent state information, you should check it for consistency.
WAIT_OBJECT_0 0x00000000L | The state of the specified object is signaled.
WAIT_TIMEOUT 0x00000102L | The time-out interval elapsed, and the object's state is nonsignaled.
WAIT_FAILED 0xFFFFFFFF | The function has failed. To get extended error information, call GetLastError.

File: include/mini/common.h (hunk 4 above)
Comment (refack) [reply include/mini/common.h:13]: We have but it makes allot of noise (Tries to find a debugger and creates a useless coredump)

File: include/mini/common.h (hunk 3 above)
Comment (refack) [reply include/mini/common.h]: So to `src/spawn-win.c`

**Comment by refack** [comment]
PTAL

File: src/spawn-win.c [hunk 7]
```diff
@@ -1,10 +1,78 @@
...
+
+static int popen_complete(const char* cmd) {
+  STARTUPINFO StartupInfo;
+  ZeroMemory(&StartupInfo, sizeof(StartupInfo));
+  StartupInfo.cb = sizeof(StartupInfo);
+
+  PROCESS_INFORMATION ProcessInformation;
+  ZeroMemory(&ProcessInformation, sizeof(ProcessInformation));
+
+/*
```
Comment (indutny) [inline src/spawn-win.c]: I'm sorry for nit-picking again. But could it be indented?

File: src/spawn-win.c (hunk 6 above)
Comment (indutny) [reply src/spawn-win.c:51]: ok. Thanks!

**Review by indutny — APPROVED** [review]
One last nit, otherwise LGTM!

File: src/spawn-win.c (hunk 7 above)
Comment (refack) [reply src/spawn-win.c]: Tried...

**Comment by indutny** [comment]
Landed with fixes in cb2b17e . Thank you!

Would you care to check that my fixes didn't break anything on Windows?

**Comment by refack** [comment]
Works. Looked at the diff, you just bunched the declarations... seems semantically equal.

**Comment by indutny** [comment]
Thanks!

**Pull request closed without merging**